app/mbedtls/platform/memcompat.c

```diff
@@ -0,0 +1,3 @@
+#include <stdlib.h>
+void *mbedtls_calloc_wrap(size_t n, size_t sz) { return calloc(n, sz); }
```

Comment:
Anything preventing us from using `calloc` and `free` directly?

Reply:
Yes, the mbedtls lib expects those two to be actual functions with standard prototypes. On the non-OS SDK they are macros which expand to `vPortXyz()` things, which makes it impossible for the mbedtls lib to do stuff like `void (*my_free)(void *) = MBEDTLS_WHATEVER_MACRO_FREE_IS;`. Hence the need for wrapper.

Reply:
Both RTOS and the SDK use the non-standard `pvPortXXXX` allocators; these are separate from the ROM-based `mem_XXXX` allocators and include extra parameters to enable some crude allocator diagnostics. 

I think that I've discussed this elsewhere, but if you look as the disassembly of these routines, they are largely wrappers around the ROM-based `mem_XXXX` allocators for most execution paths.

Reply:
@jmattsson: I see.  If there are other places in the tree where we need "actual `malloc` and `free` functions", maybe move this to `platform.c`?  But keeping them where they are is fine.

Reply:
I considered it, but I didn't want to make everyone else suffer an extra call level just because of mbedtls. Sure, it'd be a marginal expense compared to what the functions do, but when in 99% of the cases it can all be handled with a macro, I went with penalising mbedtls only. But yes, we could put actual `malloc()/free()` functions into our `app/libc/` if so wanted. We could also experiment and see how the SDK's libc would handle it if we provided the missing `_malloc_r()/_free_r()` functions (obviously without the reentrancy support) as that might allow us to not have to shadow functions it already has, but that's a whole 'nother barrel o' fish.